satpy/node.py

```diff
@@ -81,12 +81,15 @@ def copy(self, node_cache=None):
         if self.name is EMPTY_LEAF_NAME:
             return self
 
-        s = Node(self.name, self.data)
+        s = self._copy_name_and_data(node_cache)
         for c in self.children:
             c = c.copy(node_cache=node_cache)
             s.add_child(c)
         return s
 
+    def _copy_name_and_data(self, node_cache=None):
+        return Node(self.name, self.data)
```

Comment:
`node_cache` is never used here. Is that correct?

Reply:
Ah I see, it is used in the subclass. should it be used here too though?

Reply:
I don't think `.data` contains other nodes if it's not a CompositorNode

Reply:
Why does `.data` matter? It is a **node** cache, right? Also, correct me if I'm wrong, but `node_cache` is never assigned to, right?

Reply:
`.data` is different for each type of Node. CompositorNode store there the prerequisite nodes, which were not being copied before and introduced the leak we observed in #1443. 

Regarding the `node_cache`, it is used in the dependency tree's copy method at least.
https://github.com/pytroll/satpy/blob/master/satpy/dependency_tree.py#L180

Reply:
> Regarding the node_cache, it is used in the dependency tree's copy method at least.

It is *provided* to the arguments there but it is never *assigned* to, right? My understanding of that line in the dependency_tree.py module is "here is a dictionary where you should store all created nodes". Nowhere in the `node.py` code is there a line that does:

```
node_cache[dataid] = node
```

And sure, the compositor node needs copies of the `.data` sub-nodes, but as a **node** cache and as `_all_nodes` is provided from the `Scene` then shouldn't **all** nodes be cached there, not just compositor nodes?

Reply:
The way I understand it it's just a collection of existing nodes `copy` can use instead of creating new ones from scratch.

_all_nodes is been assigned to in the dep tree I think.

Reply:
Ok, then maybe `existing_nodes` might be less confusing but that's not that important. Doesn't `_all_nodes` include composite nodes as well as from-reader nodes? In which case, won't the `copy` take them from the "cache" instead of properly copying them?

Reply:
When copying a dependency tree, the first step is to create an empty tree. Then copies of the nodes of the original tree are added one by one to the new tree, populating _all_nodes along the way, ensuring that only node copies are available to the new tree.

Reply:
But within a single dependency tree, the idea is to avoid duplicates, so each node should be unique

Reply:
Ok I looked through the code a lot closer. I see now that the key to this is the `add_child` of the tree class which I'm unfamiliar with since your last refactor of this logic. I'll just assume that the cache is being added to in a useful way in `add_child`. If you think about the first iteration of the dependency tree `copy` method, the node_cache is not populated until the `add_child` at the end of the loop. That was the confusing part to me (the cache has no purpose that first iteration).

If this is ready for a full review let me know and I'll look at it next week. Then do you want me to make a Satpy release with this merged?

Reply:
Let me have a closer look tonight, but it would be nice to have it released soon yes!

Reply:
I think it's ready for review @djhoese!